File: app/src/main/kotlin/com/moviereel/ui/fragments/movie/nowplaying/MovieNPAdapter.java
package com.moviereel.ui.fragments.movie.nowplaying;

import android.content.Context;
import android.content.Intent;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.resource.drawable.GlideDrawable;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.target.Target;
import com.google.gson.Gson;
import com.moviereel.BuildConfig;
import com.moviereel.R;

import com.moviereel.data.api.model.movie.response.MovieResultsResponse;
import com.moviereel.di.ActivityContext;
import com.moviereel.di.ApplicationContext;
import com.moviereel.ui.activities.itemdetails.movie.MovieDetailsActivity;
import com.moviereel.ui.base.BaseViewHolder;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * @author lusinabrian on 01/06/17.
 * @Notes
 */

public class MovieNPAdapter extends RecyclerView.Adapter<BaseViewHolder> {
    private static final String TAG = MovieNPAdapter.class.getSimpleName();

    public static final int VIEW_TYPE_EMPTY = 0;
    public static final int VIEW_TYPE_NORMAL = 1;
    private Callback mCallback;

    private List<MovieResultsResponse> movieResultsResponseList;
    private Context mContext;

    @Inject
    public MovieNPAdapter(List<MovieResultsResponse> movieResultsResponseList, @ActivityContext Context mContext){
        this.movieResultsResponseList = movieResultsResponseList;
        this.mContext = mContext;
    }

    public void setCallback(Callback callback) {
        mCallback = callback;
    }

    @Override
    public BaseViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        switch (viewType) {
            case VIEW_TYPE_NORMAL:
                return new ViewHolder(
                        LayoutInflater.from(parent.getContext()).inflate(R.layout.item_movie_layout, parent, false));
            case VIEW_TYPE_EMPTY:
            default:
                return new EmptyViewHolder(
                        LayoutInflater.from(parent.getContext()).inflate(R.layout.item_empty_view, parent, false));
        }
    }

    @Override
    public void onBindViewHolder(BaseViewHolder baseViewHolder, int i) {
        baseViewHolder.onBind(i);
    }

    @Override
    public int getItemViewType(int position) {
        if (movieResultsResponseList != null && movieResultsResponseList.size() > 0) {
            return VIEW_TYPE_NORMAL;
        } else {
            return VIEW_TYPE_EMPTY;
        }

    }

    @Override
    public int getItemCount() {
        if(movieResultsResponseList != null && movieResultsResponseList.size() > 0){
            return movieResultsResponseList.size();
        }else {
            return 1;
        }
    }

    /**
     * Add Movie now playing responses to list
     * */
    public void addItems(List<MovieResultsResponse> movieResultsResponses){
        movieResultsResponseList.addAll(movieResultsResponses);
        notifyDataSetChanged();
    }

    public class ViewHolder extends BaseViewHolder{
        @BindView(R.id.item_movie_progressbar)
        ProgressBar progressBar;

        @BindView(R.id.item_movie_imgView)
        ImageView movieImageView;

        @BindView(R.id.item_movie_title_txtView)
        TextView movieTitle;

        @BindView(R.id.item_movie_voteAvg_txtView)
        TextView movieVoteAvg;

        @BindView(R.id.item_movieCategories_txtView)
        TextView movieCategories;

        @BindView(R.id.item_movieRuntime_txtView)
        TextView movieRuntime;

        public ViewHolder(@NotNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @Override
        protected void clear() {

        }

        public void onBind(final int position){
            final MovieResultsResponse movieResultsResponse = movieResultsResponseList.get(position);
            movieTitle.setText(movieResultsResponse.getTitle());
            movieVoteAvg.setText(String.valueOf(movieResultsResponse.getVoteAverage()));

            // glide images to image views
            Glide.with(mContext)
                    .load(BuildConfig.POSTER_PATH + movieResultsResponse.getPosterPath())
                    .listener(new RequestListener<String, GlideDrawable>() {
                        @Override
                        public boolean onException(Exception e, String model, Target<GlideDrawable> target, boolean isFirstResource) {
                            return false;
                        }

                        @Override
                        public boolean onResourceReady(GlideDrawable resource, String model, Target<GlideDrawable> target, boolean isFromMemoryCache, boolean isFirstResource) {
                            progressBar.setVisibility(View.GONE);
                            return false;
                        }
                    })
                    .fitCenter()
                    .crossFade()
                    .into(movieImageView);
            // movieRuntime.setText(movieResultsResponse.get);
            // nowPlayingResponse.getResults()
            // movieRuntime.setText(nowPlayingResponse.get);

            // open movie details activity
            itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Intent openMovieIntent = new Intent(mContext, MovieDetailsActivity.class);

                    openMovieIntent.putExtra("MovieObj", movieResultsResponse);
                    itemView.getContext().startActivity(openMovieIntent);
                }
            });
        }
    }

    public class EmptyViewHolder extends BaseViewHolder{

        @BindView(R.id.btn_retry)
        Button retryButton;

        @BindView(R.id.tv_message)
        TextView messageTextView;

        public EmptyViewHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        @Override
        protected void clear() {

        }

        @OnClick(R.id.btn_retry)
        void onRetryClick() {
            if (mCallback != null)
                mCallback.onViewEmptyViewRetryClick();
        }
    }

    public interface Callback {
        void onViewEmptyViewRetryClick();
    }

}

File: app/src/main/kotlin/com/moviereel/ui/movie/nowplaying/MovieNPAdapter.kt
package com.moviereel.ui.movie.nowplaying

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.bumptech.glide.Glide
import com.bumptech.glide.load.resource.drawable.GlideDrawable
import com.bumptech.glide.request.RequestListener
import com.bumptech.glide.request.target.Target
import com.moviereel.BuildConfig
import com.moviereel.R
import com.moviereel.data.db.entities.movie.MovieNPEntity
import com.moviereel.ui.base.BaseRecyclerAdapter
import com.moviereel.ui.base.BaseViewHolder
import kotlinx.android.synthetic.main.item_entertainment_layout.view.*
import org.jetbrains.anko.toast
import javax.inject.Inject

/**
 * @author lusinabrian on 01/06/17.
 * @Notes
 */

class MovieNPAdapter
@Inject
constructor(val movieNPEntityList: ArrayList<MovieNPEntity>) : BaseRecyclerAdapter<MovieNPEntity>(movieNPEntityList) {

    val VIEW_TYPE_LOADING = 0
    val VIEW_TYPE_NORMAL = 1

    lateinit var mCallback: Callback

    fun setCallback(callback: Callback) {
        mCallback = callback
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseViewHolder<MovieNPEntity> {
        return when (viewType) {
            VIEW_TYPE_NORMAL -> {
                val v = LayoutInflater.from(parent.context).inflate(R.layout.item_entertainment_layout, parent, false)
                ViewHolder(v)
            }
            VIEW_TYPE_LOADING -> {
                val v = LayoutInflater.from(parent.context).inflate(R.layout.progress_dialog, parent, false)
                LoadingViewHolder(v)
            }
            else -> {
                val v = LayoutInflater.from(parent.context).inflate(R.layout.item_empty_view, parent, false)
                EmptyViewHolder(v)
            }
        }
    }

    override fun getItemViewType(position: Int): Int {
        return if (movieNPEntityList.size > 0) {
            VIEW_TYPE_NORMAL
        } else {
            VIEW_TYPE_LOADING
        }

    }

    override fun getItemCount(): Int {
        return movieNPEntityList.size
    }

    inner class ViewHolder(itemView: View) : BaseViewHolder<MovieNPEntity>(itemView) {

        override fun onBind(position: Int) {
            val movieEntity = movieNPEntityList[position]

            with(itemView) {
                itemTitleTxtView.text = movieEntity.title
                itemVoteAvgTxtView.text = movieEntity.voteAverage.toString()
                // itemMovieCategoriesTxtView
                // itemMovieRuntimeTxtView
                // glide images to image views
                Glide.with(context)
                        .load(BuildConfig.POSTER_PATH + movieEntity.posterPath)
                        .listener(object : RequestListener<String, GlideDrawable> {
                            override fun onException(e: Exception, model: String, target: Target<GlideDrawable>, isFirstResource: Boolean): Boolean {
                                return false
                            }

                            override fun onResourceReady(resource: GlideDrawable, model: String, target: Target<GlideDrawable>, isFromMemoryCache: Boolean, isFirstResource: Boolean): Boolean {
                                itemProgressbar.visibility = View.GONE
                                return false
                            }
                        })
                        .fitCenter()
                        .crossFade()
                        .into(itemImgView)
                // movieRuntime.setText(movieResultsResponse.get);
                // nowPlayingResponse.getResults()
                // movieRuntime.setText(nowPlayingResponse.get);

                // open movie details activity
                setOnClickListener {
                    context.toast("$movieEntity")
//                    context.startActivity<MovieDetailsActivity>(
//                            "MovieObj" to movieEntity
//                    )
                }
            }

        }
    }

    inner class EmptyViewHolder(itemView: View) : BaseViewHolder<MovieNPEntity>(itemView) {

        fun onRetryClick() {
            // btn_retry
            mCallback.onViewEmptyViewRetryClick()
        }
    }

    inner class LoadingViewHolder(itemView: View) : BaseViewHolder<MovieNPEntity>(itemView)

    interface Callback {
        fun onViewEmptyViewRetryClick()
    }

}
